### PR TITLE
Adds determination of school of record for student demog update

### PR DESCRIPTION
**Pull request by eckermania**

### CODE REVIEW

File: api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java [hunk 1]
```diff
@@ -58,22 +69,110 @@ public void updateStudent(final Event event, final SdcSagaEntity saga, final Upd
...
+            log.debug("message sent to STUDENT_API_TOPIC for UPDATE_STUDENT Event.");
+
+        } else {
+            final Event nextEvent = Event.builder().sagaId(saga.getSagaId())
+                    .eventType(UPDATE_STUDENT)
+                    .eventOutcome(NO_STUDENT_UPDATE_NEEDED)
+                    .eventPayload(JsonUtil.getJsonStringFromObject(updateStudentSagaData))
+                    .build();
+            this.postMessageToTopic(this.getTopicToSubscribe(), nextEvent);
+            log.debug("message sent to UPDATE_STUDENT_DOWNSTREAM_SAGA_TOPIC for STUDENT_NOT_UPDATED Event.");
```
Comment (arcshiftsolutions) [inline api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java:102]: Adjust the message for the event outcome of NO_STUDENT_UPDATE_NEEDED.

File: api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java [hunk 2]
```diff
@@ -58,22 +69,110 @@ public void updateStudent(final Event event, final SdcSagaEntity saga, final Upd
...
-        final Event nextEvent = Event.builder().sagaId(saga.getSagaId())
-                .eventType(UPDATE_STUDENT)
-                .replyTo(this.getTopicToSubscribe())
-                .eventPayload(JsonUtil.getJsonStringFromObject(studentDataFromEventResponse))
-                .build();
-        this.postMessageToTopic(STUDENT_API_TOPIC.toString(), nextEvent);
-        log.debug("message sent to STUDENT_API_TOPIC for UPDATE_STUDENT Event.");
+        final List<SdcSchoolCollectionStudentEntity> otherStudentsWithSameAssignedID = sdcSchoolCollectionStudentRepository.findAllDuplicateStudentsByCollectionID(updateStudentSagaData.getCollectionID(), Collections.singletonList(updateStudentSagaData.getAssignedStudentID()));
+
+        if (updateStudentSagaData.getCollectionTypeCode().equalsIgnoreCase(CollectionTypeCodes.JULY.getTypeCode()) ||
```
Comment (arcshiftsolutions) [inline api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java:74]: I think there's a logic issue here; are we not ignoring all students in the July collection?

File: api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java [hunk 3]
```diff
@@ -58,22 +69,110 @@ public void updateStudent(final Event event, final SdcSagaEntity saga, final Upd
...
+        } else if((currStudent.getNumberOfCourses() == null && maxCourseNumberFromOtherStudents > 0) ||
+                (currStudent.getNumberOfCourses() != null && Float.parseFloat(currStudent.getNumberOfCourses()) < maxCourseNumberFromOtherStudents)){
+            return false;
+        } else {
+            Optional<SchoolTombstone> currStudentSchoolTombstone = restUtils.getSchoolByMincode(currStudent.getMincode());
+            List<Optional<SchoolTombstone>> otherStudentSchoolTombstones = retrieveSchoolTombstones(maxCourseNumberFromOtherStudents, otherStudents);
+
+            boolean otherSchoolsIncludePublic = otherStudentSchoolTombstones.stream().flatMap(Optional::stream).anyMatch(tombstone -> SchoolCategoryCodes.PUBLIC.getCode().equals(tombstone.getSchoolCategoryCode()));
+
+            if (currStudentSchoolTombstone.isPresent()){
```
Comment (arcshiftsolutions) [inline api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java:120]: Throw an exception if there's no school present - should be an impossible case.

File: api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java (hunk 2 above)
Comment (eckermania) [reply api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java:74]: Good question - I was following what was said in the ticket (https://eccbc.atlassian.net/jira/software/c/projects/EDX/boards/17?selectedIssue=EDX-2964) - "In July – do not change the school of record.". But, in hindsight, my application of this doesn't make any sense. Are we not supposed to do any updates to the student api after the july collection?

File: api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java (hunk 1 above)
Comment (eckermania) [reply api/src/main/java/ca/bc/gov/educ/studentdatacollection/api/orchestrator/UpdateStudentDownstreamOrchestrator.java:102]: yes - good catch, ty!

**Review by arcshiftsolutions — APPROVED** [review]

**Pull request merged**